### PR TITLE
feat: Typescript 1/4

### PR DESCRIPTION
**Pull request by jimmycallin**
Partially resolves FT-0403674e-a5a4-11ed-96fe-bef30e7c0f4c

- [x] I have added automatic tests where applicable
- [x] The PR title is suitable as a release note
- [x] The PR contains a description of what has been changed
- [x] The description contains manual test instructions

## Changes

Step 1 of 4 in Typescript migration. Starts with setup and minimal amount of changes for typescript to build.

## Test

Test against another codebase (preferably with typescript support) and make sure it builds and runs correctly.

### CODE REVIEW

**Review by WaylonFtrack — APPROVED** [review]

File: vite.config.ts [hunk 1]
```diff
@@ -1,13 +1,27 @@
...
+      fallbackCJS: true,
+    },
+  },
   build: {
     minify: false,
     sourcemap: true,
     lib: {
-      entry: path.resolve(__dirname, "source/index.js"),
+      entry: path.resolve(__dirname, "source/index.ts"),
       name: "ftrack-javascript-api",
```
Comment (WaylonFtrack) [inline vite.config.ts:25]: the `name` here should be something like `FtrackApi`, which is the exposed global variable.

**Review by WaylonFtrack — CHANGES_REQUESTED** [review]

File: source/error.ts [hunk 2]
```diff
@@ -12,15 +8,16 @@
...
-function errorFactory(name) {
-  function CustomError(message, errorCode) {
-    this.name = name;
-    this.message = message;
-    this.errorCode = errorCode;
-    this.stack = new Error().stack;
-  }
+function errorFactory(name: string) {
+  class CustomError extends Error {
+    errorCode: any;
```
Comment (gismya) [inline source/error.ts:13]: This seems incorrect

File: vite.config.ts (hunk 1 above)
Comment (jimmycallin) [reply vite.config.ts:25]: the `name` is only used for the build output files, which are named `ftrack-javascript-api.umd.js` or `ftrack-javascript-api.es.js`. the name of the exposed variable for the umd file is `ftrack`, which is mapped under `rollupOptions.global`

File: source/error.ts (hunk 2 above)
Comment (jimmycallin) [reply source/error.ts]: fixed!

**Review by gismya — APPROVED** [review]
Good job!

**Review by WaylonFtrack — APPROVED** [review]

**Pull request merged**